File: advent_of_code/2022/day_03/solution.py
from string import ascii_letters
from utils.py_utils.constants import INPUT
from utils.py_utils.decorators import strip_newlines, to_list


@to_list
@strip_newlines
def get_data(variation):
    with open(f"inputs/03_{variation}.txt") as f:
        data = f.readlines()
    return data


LETTERS = list(ascii_letters)

common_item = lambda el: set.intersection(*(map(set, el))).pop()
priority = lambda el: LETTERS.index(el) + 1


def easy(data):
    compartments = [[el[:len(el)//2], el[len(el)//2:]] for el in data]
    common_items = [common_item(el) for el in compartments]
    priorities = [priority(el) for el in common_items]

    return sum(priorities)


def hard(data):
    groups = [data[i:i+3] for i in range(0, len(data), 3)]
    common_items = [common_item(el) for el in groups]
    priorities = [priority(el) for el in common_items]

    return sum(priorities)


if __name__ == "__main__":
    data = get_data(INPUT)
    print(easy(data))
    print(hard(data))

File: advent_of_code/2022/day_03/solution.rb
require_relative "../utils/ruby_utils/constants"


def raw_data(variation)
  File.readlines("inputs/03_#{variation}.txt",  chomp: true)
end

LETTERS = [*('a'..'z'), *('A'..'Z')]

COMMON_ITEM = -> list { list.map(&:chars).inject(:&)[0] }
PRIORITY = -> (el) { LETTERS.find_index(el) + 1 }

def easy(data)
  compartments = data.map { |el| [el[..el.size/2], el[el.size/2..]] }
  common_items = compartments.map { |el| COMMON_ITEM.call(el) }
  priorities = common_items.map { |el| PRIORITY.call(el) }

  priorities.sum
end

def hard(data)
  groups = data.each_slice(3).to_a
  common_items = groups.map { |el| COMMON_ITEM.call(el) }
  priorities = common_items.map { |el| PRIORITY.call(el) }

  priorities.sum
end

if __FILE__ == $0
  data = raw_data(INPUT)
  puts easy(data)
  puts hard(data)
end
